modules/ingest-common/src/main/java/org/opensearch/ingest/common/FingerprintProcessor.java

```diff
@@ -0,0 +1,271 @@
+        messageDigest.update(concatenatedFields.toString().getBytes(StandardCharsets.UTF_8));
+        document.setFieldValue(targetField, Base64.getEncoder().encodeToString(messageDigest.digest()));
+
+        return document;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    /**
+     * Convert a map containing nested fields to a flattened map,
+     * for example, if the original map is
+     * {
+     *     "a": {
+     *         "b": 1,
+     *         "c": 2
+     *     }
+     * }, then the converted map is
+     * {
+     *     "a.b": 1,
+     *     "a.c": 2
+     * }
+     * @param map the original map which may contain nested fields
+     * @return a flattened map which has only one level fields
+     */
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> toFlattenedMap(Map<String, Object> map) {
+        Map<String, Object> flattenedMap = new HashMap<>();
+        for (Map.Entry<String, Object> entry : map.entrySet()) {
+            if (entry.getValue() instanceof Map) {
+                toFlattenedMap((Map<String, Object>) entry.getValue()).forEach(
+                    (key, value) -> flattenedMap.put(entry.getKey() + "." + key, value)
+                );
+            } else {
+                flattenedMap.put(entry.getKey(), entry.getValue());
+            }
+        }
+        return flattenedMap;
+    }
+
+    /**
+     * The supported hash methods used to generate hash value
+     */
+    enum HashMethod {
+        MD5(MessageDigests.md5()),
+        SHA1(MessageDigests.sha1()),
+        SHA256(MessageDigests.sha256()),
+        SHA3256(MessageDigests.sha3256());
+
+        private final MessageDigest messageDigest;
+
+        HashMethod(MessageDigest messageDigest) {
+            this.messageDigest = messageDigest;
+        }
+
+        public static MessageDigest fromMethodName(String methodName) {
+            String name = methodName.toUpperCase(Locale.ROOT);
+            switch (name) {
+                case "MD5":
+                    return MD5.messageDigest;
+                case "SHA-1":
+                    return SHA1.messageDigest;
+                case "SHA-256":
+                    return SHA256.messageDigest;
+                case "SHA3-256":
+                    return SHA3256.messageDigest;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    public static final class Factory implements Processor.Factory {
+        @Override
+        public FingerprintProcessor create(
+            Map<String, Processor.Factory> registry,
+            String processorTag,
+            String description,
+            Map<String, Object> config
+        ) throws Exception {
+            List<String> fields = ConfigurationUtils.readOptionalList(TYPE, processorTag, config, "fields");
+            List<String> excludeFields = ConfigurationUtils.readOptionalList(TYPE, processorTag, config, "exclude_fields");
```

Comment:
Could we please add validation that `fields` & `excludeFields` do not have intersection? Also, I think `Set` would be a better data structure to get rid of possible duplicates.

Reply:
Same to the [remove processor](https://github.com/opensearch-project/OpenSearch/blob/581fcd2c3204f84ff93c626be41698ec35641680/modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java#L75), fields and exclude_fields are mutually exclusive, either of them can be non-empty.

Reply:
> Same to the [remove processor](https://github.com/opensearch-project/OpenSearch/blob/581fcd2c3204f84ff93c626be41698ec35641680/modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java#L75), fields and exclude_fields are mutually exclusive, either of them can be non-empty.

This is fine, but setting `fields = ["name"]` and `excludeFields = ["name"]` looks like misconfiguration, isn't it?

Reply:
Yes, in this case, an error will be thrown: `either fields or exclude_fields can be set`, user has to configure either of `fields` and `exclude_fields`, but not both.

Reply:
Got it, thank you